aten/src/ATen/ParallelOpenMP.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+#include <ATen/ATen.h>
+
+#include <cstddef>
+#include <exception>
+
+#ifdef _OPENMP
+#define INTRA_OP_PARALLEL
+
+#include <omp.h>
+#endif
+
+namespace at {
+
+template <class F>
+inline void parallel_for(
+    const int64_t begin,
+    const int64_t end,
+    const int64_t grain_size,
+    const F& f) {
+#ifdef _OPENMP
+#pragma omp parallel if (!omp_in_parallel() && ((end - begin) >= grain_size))
+  {
+    int64_t num_threads = omp_get_num_threads();
+    int64_t tid = omp_get_thread_num();
+    int64_t chunk_size = divup((end - begin), num_threads);
+    int64_t begin_tid = begin + tid * chunk_size;
+    if (begin_tid < end) {
+      f(begin_tid, std::min(end, chunk_size + begin_tid));
+    }
+  }
+#else
+  if (begin < end) {
+    f(begin, end);
+  }
+#endif
+}
+
+template <class scalar_t, class F, class SF>
+inline scalar_t parallel_reduce(
+    const int64_t begin,
+    const int64_t end,
+    const int64_t grain_size,
+    const scalar_t ident,
+    const F& f,
+    const SF& sf) {
+  if (in_parallel_region() || get_num_threads() == 1) {
+    return f(begin, end, ident);
+  } else {
+    const int64_t num_results = divup((end - begin), grain_size);
+    std::vector<scalar_t> results(num_results);
+    scalar_t* results_data = results.data();
+#pragma omp parallel for if ((end - begin) >= grain_size)
```

Comment:
Do we not want to guard this using _OPENMP as well?

Reply:
we used `_OPENMP` above because of `omp_get_num_threads`, `omp_get_thread_num` from `omp.h` which include is guarded by `_OPENMP`